design/metadata-handling.md

```diff
@@ -0,0 +1,417 @@
+                "ethernet_mac_address": "{{ '{{ bareMetalHostMACByName "eth0" }}' }}"
+            },
+            {
+                "id": "enp2s0",
+                "type": "phy",
+                "ethernet_mac_address": "{{ '{{ bareMetalHostMACByName "eth1" }}' }}"
+            }
+        ],
+        "networks": [
+            {
+                "id": "Provisioning",
+                "type": "ipv4_dhcp",
+                "link": "enp1s0"
+            },
+            {
+                "id": "Baremetal",
+                "type": "ipv4_dhcp",
+                "link": "enp2s0",
+            }
+        ],
+        "services": [
+            {
+                "type": "dns",
+                "address": "8.8.8.8"
+            }
+        ]
+    }
+status:
+  indexes:
+    "0": "machine-1"
+  metaDatasecrets:
+    "machine-1":
+        name: machine-1-metadata-0
+        namespace: default
+  networkDatasecrets:
+    "machine-1":
+        name: machine-1-networkdata-0
+        namespace: default
+  lastUpdated: "2020-03-25T12:33:25Z"
+```
+
+This object will be reconciled by its own controller. When reconciled,
+the controller will add a label pointing to the Metal3Cluster that has nodes
+linking to this object. The spec contains a `metaData` and a `networkData` field
+that contain a template of the values that will be rendered for all nodes.
+
+The `metaData` field should contain a map of strings in yaml format, while
+`networkData` should contain a json string that fulfills the requirements of
+[Nova network_data.json](https://docs.openstack.org/nova/latest/user/metadata.html#openstack-format-metadata).
+The format definition can be found
+[here](https://docs.openstack.org/nova/latest/_downloads/9119ca7ac90aa2990e762c08baea3a36/network_data.json).
+
+The values will be [go templates](
+https://golang.org/pkg/text/template/). Multiple functions would be available :
+
+- **machineName** : returns the Machine name
+- **metal3MachineName** : returns the Metal3Machine name
+- **bareMetalHostName** : returns the BareMetalHost name
+- **index** : returns the Metal3Machine index for the Metal3Metadata object.
+  The index starts from 0.
+- **indexWithOffset** : takes an integer as parameter and returns the sum of
+  the index and the offset parameter
+- **indexWithStep** : takes an integer as parameter and returns the
+  multiplication of the index and the step parameter
+- **indexWithOffsetAndStep** OR **indexWithStepAndOffset**: takes two
+  integers as parameters, order depending on the function name, and returns the
+  sum of the offset and the multiplication of the index and the step.
+- **index*Hex** : All the `index` functions can be suffixed with `Hex` to
+  get the same value in hexadecimal format.
+- **bareMetalHostMACByName**: takes a string as parameter and returns the MAC
+  address of the nic with the name matching the parameter. This function
+  operates over the list of NICs in the `status.hardwareDetails.NIC` field of
+  the BareMetalHost.
```

Comment:
How does the thing rendering the template associated with the Metal3Machine know which BareMetalHost to use?

Reply:
the data is rendered when the BMH is associated, so the controller can access all the fields from the Metal3Machine or the BMH